### PR TITLE
Add support for transposed convolution negative input padding

### PR DESCRIPTION
**Pull request by ivangarcia44**
Currently when a transposed convolution is lowered from the torch dialect to the linalg dialect we get an insert_slide operation to create padding for the input tensor. For example:

%inserted_slice = tensor.insert_slice %arg0 into %cast[0, 0, 2, %c-1] [1, 1, 4, 7] [1, 1, 1, 1] : tensor<1x1x4x7xf32> into tensor<1x1x?x?xf32>

The above works well for the case where the input padding is positive. For transposed convolution the input padding is defined with this formula: dilation * (kernel_size - 1) - padding (see https://pytorch.org/docs/stable/generated/torch.nn.ConvTranspose2d.html) for details.

Notice that if the input padding is above the left hand term, we get negative padding. For these cases PyTorch reduces the size of the input tensor. The torch to linalg lowering was not doing this, and therefore its value does not match what PyTorch gives (captured in e2e tests TransposedConv2dNegativePadding and TransposedConv3dNegativePadding).

To fix this a tensor.extract_slice operation is added just before the insert_slice operation to reduce the input tensor size as PyTorch does. In the example above we get the code below whose result matches the numerical values of PyTorch.

    %extracted_slice = tensor.extract_slice %arg0[0, 0, 0, 1] [1, 1, 4, 5] [1, 1, 1, 1] : tensor<1x1x4x7xf32> to tensor<1x1x4x5xf32>
    %inserted_slice = tensor.insert_slice %extracted_slice into %4[0, 0, 2, 0] [1, 1, 4, 5] [1, 1, 1, 1] : tensor<1x1x4x5xf32> into tensor<1x1x8x5xf32>

For each dimension with a negative padding, we add a positive offset (absolute value of negative padding) in the corresponding dimension for the extract_slice operation, and the dimension size is reduced by twice that amount (elements are lost in both sides of the dimension as specified in PyTorch).

Then on the insert_slice the negative padding dimension has an offset of zero because the trimmed dimension fits exactly. For the case when padding is positive the existing behavior is kept.

@rsuderman
@vivekkhandelwal1
@zjgarvey
@penguin-wwy
@ubfx
@sahas3
@Hanumanth04 
@dixinzhou
@rafaelubalmw

### CODE REVIEW

File: test/Conversion/TorchToLinalg/convolution.mlir [hunk 1]
```diff
@@ -150,3 +150,27 @@ func.func @transposedGroupedConvolution2D(%arg0: !torch.vtensor<[1,2,5,7],f32>)
...
   return %6 : !torch.vtensor<[1,4,10,14],f32>
 }
+
+// CHECK-LABEL:   func.func @tranConv2dNegativePadding(
+// CHECK-SAME:       %[[INPUT_VTENSOR:.*]]: !torch.vtensor<[1,1,4,7],f32>) -> !torch.vtensor<[1,2,6,3],f32>
+// CHECK:            %[[IN_TENSOR:.*]] = torch_c.to_builtin_tensor %[[INPUT_VTENSOR:.*]] : !torch.vtensor<[1,1,4,7],f32> -> tensor<1x1x4x7xf32>
+// CHECK:            %[[EXTRACTED_SLICE:.*]] = tensor.extract_slice %[[IN_TENSOR:.*]][0, 0, 0, 1] [1, 1, 4, 5] [1, 1, 1, 1] : tensor<1x1x4x7xf32> to tensor<1x1x4x5xf32>
+// CHECK:            %[[INSERTED_SLICE:.*]] = tensor.insert_slice %[[EXTRACTED_SLICE:.*]] into %[[INIT_TENSOR:.*]][0, 0, 2, 0] [1, 1, 4, 5] [1, 1, 1, 1] : tensor<1x1x4x5xf32> into tensor<1x1x8x5xf32>
+// CHECK:            %[[OUT_TENSOR:.*]] = linalg.conv_2d_nchw_fchw {dilations = dense<1> : vector<2xi64>, strides = dense<1> : vector<2xi64>} ins(%[[INSERTED_SLICE:.*]], %[[WEIGHTS:.*]] : tensor<1x1x8x5xf32>, tensor<2x1x3x3xf32>) outs(%[[INIT_OUT_TENSOR:.*]] : tensor<1x2x6x3xf32>) -> tensor<1x2x6x3xf32>
+// CHECK:            %[[OUT_VTENSOR:.*]] = torch_c.from_builtin_tensor %[[OUT_TENSOR:.*]] : tensor<1x2x6x3xf32> -> !torch.vtensor<[1,2,6,3],f32>
```
Comment (zjgarvey) [inline test/Conversion/TorchToLinalg/convolution.mlir:160]: Don't add the `:*` on values  that should match previous definitions, otherwise it will just overwrite the variables.

**Review by zjgarvey — COMMENTED** [review]
Nice! This looks good to me for the most part. Small change on the lit test and let's get this merged.

File: test/Conversion/TorchToLinalg/convolution.mlir (hunk 1 above)
Comment (ivangarcia44) [reply test/Conversion/TorchToLinalg/convolution.mlir]: Done, updated the change. Thanks

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py [hunk 2]
```diff
@@ -1757,3 +1757,67 @@ def forward(self, inputVec, weight, bias):
 @register_test_case(module_factory=lambda: ConvolutionModule2DGroupedTranspose())
 def ConvolutionModule2DGroupedTranspose_basic(module, tu: TestUtils):
     module.forward(tu.rand(1, 2, 5, 7), tu.rand(2, 2, 3, 3), tu.rand(4))
+
+
+class TransposedConv2dNegativePadding(torch.nn.Module):
```
Comment (sahas3) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/conv.py:1762]: Should there be a 1d test too?

**Review by sahas3 — APPROVED** [review]
LGTM, just a question about adding a 1d test.
